File: rust/543.diameter-of-binary-tree.rs
/*
 * @lc app=leetcode id=543 lang=rust
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut length = 0;
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, length: &mut i32) -> i32 {
            if let Some(n) = node {
                let nb = n.borrow();
                let left_length = dfs(&nb.left, length);
                let right_length = dfs(&nb.right, length);
                *length = i32::max(*length, left_length + right_length);
                return 1 + left_length.max(right_length);
            }
            0
        }
        dfs(&root, &mut length);
        length
    }
}
// @lc code=end

File: javascript/543.diameter-of-binary-tree.js
/*
 * @lc app=leetcode id=543 lang=javascript
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  let d = 0;
  function dfs(node) {
    if (node) {
      let l = dfs(node.left);
      let r = dfs(node.right);
      d = Math.max(d, l + r);
      return 1 + Math.max(l, r);
    }
    return 0;
  }
  dfs(root);
  return d;
};
// @lc code=end
